tensortrade/stochastic/processes/gan.py

```diff
@@ -0,0 +1,381 @@
+        self._num_generated_features = kwargs.get('num_generated_features', 1)
+        self._output_shape = kwargs.get('output_shape', (self._times_to_generate, 5, 1))
+        self._initialize_gan()
+        self._train_gan()
+        self.generate_price_history()
+
+    def _initialize_placeholders(self):
+        self._CG = tf.placeholder(tf.float32,
+                                  [self._batch_size, self._seq_length])
+        self._CD = tf.placeholder(tf.float32, [self._batch_size, self._seq_length])
+        self._Z = tf.placeholder(tf.float32, [self._batch_size, self._seq_length, self._latent_dim])
+        self._W_out_G = tf.Variable(tf.truncated_normal([self._hidden_units_g, self._num_generated_features]))
+        self._b_out_G = tf.Variable(tf.truncated_normal([self._num_generated_features]))
+
+        self._X = tf.placeholder(tf.float32, [self._batch_size, self._seq_length, self._num_generated_features])
+        self._W_out_D = tf.Variable(tf.truncated_normal([self._hidden_units_d, 1]))
+        self._b_out_D = tf.Variable(tf.truncated_normal([1]))
+
+    @staticmethod
+    def sample_Z(batch_size, seq_length, latent_dim):
+        # --- to do with latent space --- #
+        sample = np.float32(np.random.normal(size=[batch_size, seq_length, latent_dim]))
+        return sample
+
+    def _cal_cagr_n_vol(self):
+        days = (self._training_data.index[-1] - self._training_data.index[0]).days
+        self._cagr = (((self._training_data['Close'][-1]) / self._training_data['Close'][1]) ** (252.0 / days)) - 1
+
+        # create a series of percentage returns and calculate the annual volatility of returns
+        self._training_data['Returns'] = self._training_data['Close'].pct_change()
+        self._volatility = self._training_data['Returns'].std() * np.sqrt(self._t)
+
+    def sample_data(self):
+        """
+        we will do this by usual Monte Carlo simulation to look at the potential evolution of asset prices
+        over time, assuming they are subject to daily returns that follow a normal distribution.
+        To set up our simulation, we need to estimate the expected level of return (mu) and volatility (vol) of the stock
+        """
+        vectors = []
+        for i in range(self._n_samples):
+            # create list of daily returns using random normal distribution
+            daily_returns = np.random.normal(self._cagr / self._t, self._volatility / np.sqrt(self._t),
+                                             self._seq_length)
+            vectors.append(np.log(daily_returns + np.absolute(min(daily_returns)) + 1))
+
+        dataset = np.array(vectors)
+        dataset.reshape(-1, self._seq_length)
+
+        return dataset
+
+    @staticmethod
+    def get_batch(samples, batch_size, batch_idx):
+        start_pos = batch_idx * batch_size
+        end_pos = start_pos + batch_size
+        return samples[start_pos:end_pos]
+
+    def generator(self, z, c):
```

Comment:
Can we have this more generalized so that if I have a custom generator and a custom discriminator I can just give those as input to this class and it can train them?